### PR TITLE
KAFKA-2377: Add basic system test for copycat using source and sink file connectors.

### PR DESCRIPTION
**Pull request by ewencp**
Tests standalone mode by running separate source and sink connectors, catting
data into the source file, and validating the output in the sink file. Restarts
the service to verify that clean restarts will result in tasks resuming where
they left off.

### CODE REVIEW

**Comment by ewencp** [comment]
FYI, this and KAFKA-2439 / #148 require an updated ducktape which requires changes to other tests (e.g. the console consumer sanity check test). I didn't make all those changes, we'd probably need to apply 2439 first, then this patch.

File: copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceTask.java [hunk 1]
```diff
@@ -65,21 +75,17 @@ public void start(Properties props) {
                             throw new CopycatException(e);
                         }
                     }
-                    log.debug("Skipped to offset {}", lastRecordedOffset);
+                    log.debug("Skipped to offset", lastRecordedOffset);
```
Comment (Ishiihara) [inline copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceTask.java]: Missing {} here?

File: copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceTask.java [hunk 2]
```diff
@@ -68,18 +78,14 @@ public void start(Properties props) {
                     log.debug("Skipped to offset {}", lastRecordedOffset);
                 }
                 streamOffset = (lastRecordedOffset != null) ? lastRecordedOffset : 0L;
+                reader = new BufferedReader(new InputStreamReader(stream));
             } catch (FileNotFoundException e) {
-                throw new CopycatException("Couldn't find file for FileStreamSourceTask: {}", e);
+                log.warn("Couldn't find file for FileStreamSourceTask, sleeping to wait for it to be created");
+                Thread.sleep(1000);
```
Comment (granders) [inline copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceTask.java]: Should this sort of "backoff" be the responsibility of whoever is polling for data?

File: tests/kafkatest/services/copycat.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+        node.account.ssh("rm -f /mnt/copycat.pid", allow_fail=False)
+
+    def restart(self):
+        # We don't want to do any clean up here, just restart the process
+        self.stop_node(self.node)
+        self.start_node(self.node)
+
+    def clean_node(self, node):
+        node.account.ssh("rm -rf /mnt/copycat.pid /mnt/copycat.log /mnt/copycat.properties /mnt/copycat-connector.properties " + " ".join(self.files), allow_fail=False)
```
Comment (granders) [inline tests/kafkatest/services/copycat.py:95]: If you're on-board with process cleanup here, probably add a `node.account.kill_process("java", clean_shutdown=False, allow_fail=True)`

File: tests/kafkatest/services/copycat.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        super(CopycatStandaloneService, self).start()
+
+    def start_node(self, node):
+        node.account.create_file("/mnt/copycat.properties", self.config_template)
+        node.account.create_file("/mnt/copycat-connector.properties", self.connector_config_template)
+
+        self.logger.info("Starting Copycat standalone process")
+        with monitor_log(self.node, "/mnt/copycat.log") as monitor:
+            node.account.ssh("/opt/kafka/bin/copycat-standalone.sh /mnt/copycat.properties /mnt/copycat-connector.properties " +
+                             "1>> /mnt/copycat.log 2>> /mnt/copycat.log & echo $! > /mnt/copycat.pid")
```
Comment (granders) [inline tests/kafkatest/services/copycat.py:60]: Having used the pid file approach in kafka service for a little while, I don't think it has any advantages over grepping the output of `ps`, and it takes more care to keep the pid file in sync

File: tests/kafkatest/services/util.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (granders) [inline tests/kafkatest/services/util.py]: Doesn't have to happen now, but this is probably useful and general enough to graduate to ducktape

File: tests/kafkatest/services/util.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    See LogMonitor for more usage information.
+    """
+    try:
+        offset = int(ssh_output(node.account, "wc -c %s" % log).split()[0])
+    except subprocess.CalledProcessError:
+        offset = 0
+    yield LogMonitor(node, log, offset)
+
+class LogMonitor(object):
```
Comment (granders) [inline tests/kafkatest/services/util.py]: This is cool and surprisingly simple.
Is there any reason why you're doing a byte-based vs line-based approach as far as "offsets" are concerned?

File: tests/kafkatest/services/copycat.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            return []
+
+    def stop_node(self, node, clean_shutdown=True):
+        pids = self.pids()
+        sig = signal.SIGTERM if clean_shutdown else signal.SIGKILL
+
+        for pid in pids:
+            self.node.account.signal(pid, sig, allow_fail=False)
+        for pid in pids:
+            wait_until(lambda: self.node.account.ssh("kill -0 " + str(pid), allow_fail=True) == 0, timeout_sec=10, err_msg="Copycat standalone process took too long to exit")
```
Comment (granders) [inline tests/kafkatest/services/copycat.py]: When I tried running, this timed out while waiting for the process to go away, when clearly the pid was actually gone (e.g. `bash: line 0: kill: (4449) - No such process`)

Shouldn't it be the opposite condition: `kill -0 <pid>` == 0 iff the process exists right?
In any case, you could also try using `node.account.alive(pid)` (see ducktape RemoteAccount.alive)

File: tests/kafkatest/tests/copycat_test.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    def test_file_source_and_sink(self):
+        # These need to be set
+        self.source.set_configs(self.render("copycat-standalone.properties"), self.render("copycat-file-source.properties"))
+        self.sink.set_configs(self.render("copycat-standalone.properties"), self.render("copycat-file-sink.properties"))
+
+        self.source.start()
+        self.sink.start()
+
+        # Generating data on the source node should generate new records and create new output on the sink node
+        self.source.node.account.ssh("echo -e -n " + repr(self.FIRST_INPUT) + " >> " + self.INPUT_FILE)
```
Comment (granders) [inline tests/kafkatest/tests/copycat_test.py:53]: This seems to be hanging as if it's gotten an unterminated quote when I run it
The actual command that is run is:
`'echo -e -n '\''foo\nbar\nbaz\n'\'' >> /mnt/copycat.input'`

**Pull request closed without merging**